ompi/mca/coll/xhc/coll_xhc.c

```diff
@@ -9,609 +9,677 @@
+    ompi_communicator_t *comm, XHC_COLLTYPE_T colltype);
 
-static void *xhc_shmem_create(opal_shmem_ds_t *seg_ds, size_t size,
-    ompi_communicator_t *ompi_comm, const char *name_chr_s, int name_chr_i);
-static void *xhc_shmem_attach(opal_shmem_ds_t *seg_ds);
 static mca_smsc_endpoint_t *xhc_smsc_ep(xhc_peer_info_t *peer_info);
 
 // ------------------------------------------------
 
 int mca_coll_xhc_lazy_init(xhc_module_t *module, ompi_communicator_t *comm) {
+    xhc_peer_info_t *peer_info = NULL;
 
     int comm_size = ompi_comm_size(comm);
     int rank = ompi_comm_rank(comm);
 
-    xhc_peer_info_t *peer_info = module->peer_info;
-
-    opal_shmem_ds_t *peer_cico_ds = NULL;
-    xhc_data_t *data = NULL;
-
-    xhc_coll_fns_t xhc_fns;
-
-    int return_code = OMPI_SUCCESS;
-    int ret;
+    int err, return_code = OMPI_SUCCESS;
 
     errno = 0;
 
-    // ----
+    if(module->init)
+        return OMPI_SUCCESS;
+    else if(module->error)
+        return OMPI_ERROR;
 
-    /* XHC requires rank communication during its initialization.
-     * Temporarily apply the saved fallback collective modules,
-     * and restore XHC's after initialization is done. */
-    xhc_module_install_fallback_fns(module, comm, &xhc_fns);
+    // ---
 
-    // ----
+    peer_info = calloc(comm_size, sizeof(xhc_peer_info_t));
+    if(!peer_info) RETURN_WITH_ERROR(return_code, OMPI_ERR_OUT_OF_RESOURCE, end);
 
-    ret = xhc_module_prepare_hierarchy(module, comm);
-    if(ret != OMPI_SUCCESS) {
-        RETURN_WITH_ERROR(return_code, ret, end);
+    for(int r = 0; r < comm_size; r++) {
+        peer_info[r].proc = ompi_comm_peer_lookup(comm, r);
+        peer_info[r].locality = peer_info[r].proc->super.proc_flags;
     }
 
-    // ----
+    peer_info[rank].locality |= ((1 << XHC_LOC_EXT_BITS) - 1) << XHC_LOC_EXT_START;
 
-    data = malloc(sizeof(xhc_data_t));
-    peer_cico_ds = malloc(comm_size * sizeof(opal_shmem_ds_t));
-    if(!data || !peer_cico_ds) {
-        RETURN_WITH_ERROR(return_code, OMPI_ERR_OUT_OF_RESOURCE, end);
-    }
+    // ---
 
-    *data = (xhc_data_t) {
-        .comms = NULL,
-        .comm_count = -1,
+    OBJ_CONSTRUCT(&module->hierarchy_cache, opal_hash_table_t);
+    err = opal_hash_table_init(&module->hierarchy_cache, XHC_COLLCOUNT);
+    if(err != OPAL_SUCCESS) RETURN_WITH_ERROR(return_code, err, end);
 
-        .pvt_coll_seq = 0
-    };
+    // ---
 
-    // ----
+    module->comm_size = comm_size;
+    module->rank = rank;
 
-    if(OMPI_XHC_CICO_MAX > 0) {
-        opal_shmem_ds_t cico_ds;
+    module->peer_info = peer_info;
+    module->init = true;
+
+    if((mca_coll_xhc_component.print_info & XHC_PRINT_INFO_CONFIG) && rank == 0) {
+        err = xhc_print_config_info(module, comm);
 
-        void *my_cico = xhc_shmem_create(&cico_ds,
-            OMPI_XHC_CICO_MAX, comm, "cico", 0);
-        if(!my_cico) {
-            RETURN_WITH_ERROR(return_code, OMPI_ERR_OUT_OF_RESOURCE, end);
+        if(err != OMPI_SUCCESS) {
```

Comment:
Minor: the Open MPI coding guidelines suggest that you use the constant in a comparison as a first argument, i.e. this would be

```
if (OMPI_SUCCESS != err) {
```

(this occurs at multiple locations in the code, will not point it out beyond this one)

Reply:
Hmm yes, I'm not too proud of my non-compliance here. But this rule is really hard to follow. The equality check is not just that, the code tells a story, and it tells it left-to-right (at least to me). So when I see `x == 0` I read it as if `x` has the value of 0, whereas in `0 == x`, `0` has the value of `x`?? And this is kind of a simple case, it can get more complex. Furthermore, this rule often leads to reversing the operands in relational operations, e.g. things like `0 < count`, where all hell breaks loose trying to read it. And as for the safety aspect, GCC has `-Wparentheses` which warns about this problem; we could even make this an error (in XHC? in all of OMPI?).

Of course this is not the platform to argue about the coding style, sorry for the manifesto :-). I can change all of the code to reverse the operands, but honestly I'm more worried about bugs creeping in as a result of this conversion, rather than an accidental omission of an `=` or `!`. But if it must be...

Reply:
Reason for this rule is very simple and compelling: if you make a typo and have `x = 0`, then that is a valid assignment and will always equate to true. However, if you have `0 = x`, that is an invalid assignment and will generate a compile error. So it is a simple and effective way to protect against mistakes.

Reply:
Sure I understand that part. IMHO it probably still doesn't outweigh the readability problem. But either way, isn't this typo nowadays easily caught with `-Wparentheses` (or even `-Werror=parentheses`)? Is there still a reason to do it manually, given the readability disadvantage of `0 = x`?

Reply:
Well, readability is a pretty subjective thing and we may not all agree with your statement. I don't know when that warning showed up, but definitely not in older compilers and perhaps in some we still support (we just finally turned off gcc6!), and I don't know if other compiler families support it. Besides, there is a tendency over here for people to ignore warnings and to strongly resist treating warnings as errors, so relying on a compiler warning to protect us is probably a losing proposition.

Given how many times this simple rule has saved our tails, it is rather difficult to justify removing it. Still, you are welcome to try and convince the community to do so!

Reply:
That's fair enough. Personally, I'd rather rely on the compiler for this, especially given the alternative, and given that in the XHC codebase I don't leave warnings around.

For some completeness, since we discussed it, a quick test reveals that the warning is present at least as far back as GCC 4.8 with `-Wall`, and it's also on the latest Clang (by default). It's also present on some ICC I tested, but only with `-Wparentheses`, not with just `-Wall`. Of course this is not exhaustive, and indeed this is not the appropriate place to figure it out either...